File: python3/node.py
from __future__ import annotations
from typing import List, Dict, Any, Union
from collections import deque

JSONDict = Dict[str, Any]

class Node:
    """
    Representation of one step into a JSON Tree
    """

    def __init__(self, json_data: Any, tree: 'JsonTree', linked_list: deque = deque([]), prior_keys: List[Union[str, int]] = []) -> None:
        self.json_data = json_data
        self.tree = tree
        self.linked_list = linked_list
        self.prior_keys = prior_keys

        self.dtype = type(self.json_data)

        self.nodes = []

        #If the node is a leaf then it has no edges
        if self.is_leaf:
            self.json_data = {prior_keys[-1]: self.json_data}
            self.tree.leaf_nodes.append(self)

        else:
            self.get_edges()

    @property
    def is_leaf(self):
        """
        If the dtype of self.json_data is not a dict or a list then it must be
        a leaf node
        """
        return self.dtype is not list and self.dtype is not dict

    def get_edges(self):
        """
        Get all edges connected to current Node
        """
        iter_arr = zip(range(len(self.json_data)),
                       self.json_data) if self.dtype is list else self.json_data.items()

        for key, value in iter_arr:
            next_linked_list = self.linked_list + deque([self])
            next_key = self.prior_keys + [key]
            node = Node(value, self.tree, next_linked_list, next_key)
            self.nodes.append(node)

    def __repr__(self):
        return str(self.json_data)

File: javascript/node.js
class Node {
    constructor(jsonData, tree, linkedList=[], priorKeys=[]) {
        this.jsonData = jsonData;
        this.tree = tree;
        this.linkedList = linkedList;
        this.priorKeys = priorKeys;

        this.dtype = this._determineObjectType(this.jsonData);
        // console.log(this.dtype);
        this.nodes = []
        
        if (this.isLeaf) {
            var lastKey = this.priorKeys[this.priorKeys.length-1];
            var lastValue = this.jsonData;
            this.jsonData = {};
            this.jsonData[lastKey] = lastValue;
            this.tree.leafNodes.push(this);
        } else {
            this._getEdges()
        }
    }

    _determineObjectType(jsonData) {
        //determine if the object passed is an array
        if (typeof(jsonData) == "object") {
            if (Object.prototype.toString.call(jsonData) == "[object Object]") {
                var dtype = 'object';
            } else {
                var dtype = 'array';
            }
        } else {
            dtype = typeof(jsonData);
        }
        return dtype;
    }

    get isLeaf() {
        //Determine whether the node is a leaf node
        return ((this.dtype != 'object') && (this.dtype != 'array'));
    }

    _getEdges() {
        //Get edges to other nodes in tree recursively
        if (this.dtype == 'array') {
            for (i = 0; i < this.jsonData.length; i++) {
                var currentData = this.jsonData[i];

                //Construct references for the next Node
                var nextLinkedList = [...this.linkedList];
                var nextPriorKeys = [...this.priorKeys];
                nextLinkedList.push(this);
                nextPriorKeys.push(i);

                var newNode = new Node(currentData, this.tree, nextLinkedList, nextPriorKeys);
                this.nodes.push(newNode);
            }
        } else {
            for (var key of Object.keys(this.jsonData)) {
                var currentData = this.jsonData[key];

                //Construct references for the next Node
                var nextLinkedList = [...this.linkedList];
                var nextPriorKeys = [...this.priorKeys];
                nextLinkedList.push(this);
                nextPriorKeys.push(key);

                var newNode = new Node(currentData, this.tree, nextLinkedList, nextPriorKeys);
                this.nodes.push(newNode);
            }
        }
    }

    _getIterArr() {
        //Return an array that we will iterate over determined by this.dtype
        if (this.dtype == 'array') {

        }
    }
}

module.exports = Node;